bin/ch/WScriptJsrt.cpp

```diff
@@ -1073,6 +1074,114 @@ JsValueRef __stdcall WScriptJsrt::LoadTextFileCallback(JsValueRef callee, bool i
     return returnValue;
 }
 
+int JsFgets(char* buf, int size, FILE* file)
+{
+    int n = size - 1;
+    if (n < 0)
+        return -1;
+
+    bool crflag = false;
+    int c, i = 0;
+
+    for (i = 0; i < n && (c = getc(file)) != EOF; i++) {
+        buf[i] = (char)c;
+        if (c == '\n') {        // any \n ends a line
+            i++;                // keep the \n; we know there is room for \0
+            break;
+        }
+        if (crflag) {           // \r not followed by \n ends line at the \r
+            ungetc(c, file);
+            break;              // and overwrite c in buf with \0
+        }
+        crflag = (c == '\r');
+    }
+
+    buf[i] = '\0';
+    return i;
+}
+
+JsValueRef __stdcall WScriptJsrt::ReadLineStdinCallback(JsValueRef callee, bool isConstructCall, JsValueRef *arguments, unsigned short argumentCount, void *callbackState)
+{
+    HRESULT hr = E_FAIL;
+    JsValueRef returnValue = JS_INVALID_REFERENCE;
+    JsErrorCode errorCode = JsNoError;
+
+    const int BUFSIZE = 256;
+    FILE* from = stdin;
+    int buflength = 0;
+    int bufsize = BUFSIZE;
+    char* buf = (char*)malloc(bufsize);
+    bool sawNewline = false;
+    char* tmp;
+    if (!buf)
+        goto Error;
+
+    int gotlength;
+    while ((gotlength = JsFgets(buf + buflength, bufsize - buflength, from)) > 0) {
+        buflength += gotlength;
+
+        /* Are we done? */
+        if (buf[buflength - 2] == '\r' && buf[buflength - 1] == '\n') {
+            buf[buflength - 1] = '\0';
+            buf[buflength - 2] = '\0';
+            sawNewline = true;
```

Comment:
Here you take two characters off the buffer, but sawNewline only indicates that you should take one character off the buffer on L1173. Why not just modify buflength here rather than trying to keep a flag alive for this?